### PR TITLE
feat: Add snooze, reopen option to bulk actions

### LINKED ISSUE #4799: Bulk actions - Followup
### Enhancements

- [x] Add Snooze, reopen option to bulk actions.
- [x] Remove the resolve option if all selected conversations are resolved
- [x] Remove the reopen option if all selected conversations are in the open state
- [x] Remove the snoozed option if all selected conversations are in snoozed state

### Bugs

- [ ] If the conversations are selected in a random order, "Select all" will not be active even if all conversations available are selected.
- [x] Checkbox selection sometime leads to the selection of the conversation.

### PR DESCRIPTION
**Pull request by fayazara**
## Description
Demo
https://www.loom.com/share/c8384f20679445508a496be68985d4bb

Fixes #4799 

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/javascript/dashboard/components/ChatList.vue [hunk 1]
```diff
@@ -364,13 +370,37 @@ export default {
...
+        return this.$store.getters.getConversationById(item).status === 'open';
+      });
+    },
+    allSelectedConversationsAreResolved() {
+      if (!this.selectedConversations.length) return false;
+      return this.selectedConversations.every(item => {
+        return (
+          this.$store.getters.getConversationById(item).status === 'resolved'
+        );
+      });
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ChatList.vue:394]: This logic is repeated. Why don't we just pass the status as a param and make this a function?

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/UpdateActions.vue [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+<style scoped lang="scss">
+.flex-between {
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+}
+.actions-container {
+  transform-origin: top right;
+  position: absolute;
+  top: 48px;
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/UpdateActions.vue:130]: Can we use rem or spacing variables here?

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/Index.vue [hunk 3]
```diff
@@ -71,22 +86,35 @@ export default {
       type: Array,
       default: () => [],
     },
+    allSelectedConversationsAreOpen: {
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/Index.vue:89]: displayOpenAction or showOpenAction?

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/AgentSelector.vue [hunk 4]
```diff
@@ -1,5 +1,16 @@
...
+    <div class="triangle">
+      <svg height="12" viewBox="0 0 24 12" width="24">
+        <path
+          d="M20 12l-8-8-12 12"
+          fill="var(--white)"
+          fill-rule="evenodd"
+          stroke="var(--s-50)"
+          stroke-width="1px"
+        ></path>
+      </svg>
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/AgentSelector.vue:12]: Let us use this SVG via fluent-icon pack

File: app/javascript/dashboard/i18n/locale/en/bulkActions.json [hunk 5]
```diff
@@ -6,12 +6,16 @@
...
         "ASSIGN_SUCCESFUL": "Conversations assigned successfully",
         "ASSIGN_FAILED": "Failed to assign conversations, please try again",
         "RESOLVE_SUCCESFUL": "Conversations resolved successfully",
         "RESOLVE_FAILED": "Failed to resolve conversations, please try again",
         "ALL_CONVERSATIONS_SELECTED_ALERT": "Conversations visible on this page are only selected.",
-        "AGENT_LIST_LOADING": "Loading Agents"
+        "AGENT_LIST_LOADING": "Loading Agents",
+        "UPDATE": {
+            "UPDATE_CONVERSATIONS_LABEL": "Update Conversations",
+            "UPDATE_SUCCESFUL": "Conversations updated successfully",
```
Comment (pranavrajs) [inline app/javascript/dashboard/i18n/locale/en/bulkActions.json:17]: Conversation status updated successfully.

**Review by pranavrajs — CHANGES_REQUESTED** [review]
I have some minor comments, otherwise a good update. Nice work!

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/AgentSelector.vue (hunk 4 above)
Comment (fayazara) [reply app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/AgentSelector.vue:12]: This not an icon, its actually the caret/triangle for showing above popover which I custom made, also this has some extra attributes like stroke-width and stroke.

Should I add it icons, or I can make a separate component for this, `woot-popover`? 

@pranavrajs

Comment (pranavrajs) [reply app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/AgentSelector.vue]: This is fine.

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/UpdateActions.vue [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  <div v-on-clickaway="onClose" class="actions-container">
+    <div class="triangle">
+      <svg height="12" viewBox="0 0 24 12" width="24">
+        <path
+          d="M20 12l-8-8-12 12"
+          fill="var(--white)"
+          fill-rule="evenodd"
+          stroke="var(--s-50)"
+          stroke-width="1px"
+        ></path>
```
[the remark this answers is not in the document]
Comment (pranavrajs) [reply app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/UpdateActions.vue]: Let us build a generic component for popover as a separate ticket

File: app/javascript/dashboard/components/ChatList.vue [hunk 7]
```diff
@@ -84,7 +84,19 @@
...
+      :conversations="selectedConversations"
+      :all-conversations-selected="allConversationsSelected"
+      :selected-inboxes="uniqueInboxes"
+      :show-open-action="allSelectedConversationsStatus('open')"
+      :show-resolved-action="allSelectedConversationsStatus('resolved')"
+      :show-snoozed-action="allSelectedConversationsStatus('snoozed')"
+      @select-all-conversations="selectAllConversations"
+      @assign-agent="onAssignAgent"
+      @update-conversations="onUpdateConversations"
+      @resolve-conversations="onResolveConversations"
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ChatList.vue:98]: This action is not used anymore.

File: app/javascript/dashboard/components/ChatList.vue [hunk 8]
```diff
@@ -592,6 +599,21 @@ export default {
...
+            status,
+          },
+        });
+        this.selectedConversations = [];
+        this.showAlert(this.$t('BULK_ACTION.UPDATE.UPDATE_SUCCESFUL'));
+      } catch (err) {
+        this.showAlert(this.$t('BULK_ACTION.UPDATE.UPDATE_FAILED'));
+      }
+    },
     async onResolveConversations() {
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/ChatList.vue:617]: Please remove this function if not used.

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/Index.vue [hunk 9]
```diff
@@ -71,40 +86,51 @@ export default {
...
     selectAll(e) {
       this.$emit('select-all-conversations', e.target.checked);
     },
     submit(agent) {
       this.$emit('assign-agent', agent);
     },
+    updateConversations(status) {
+      this.$emit('update-conversations', status);
+    },
     resolveConversations() {
```
Comment (pranavrajs) [inline app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/Index.vue:118]: This method is not used.

**Review by pranavrajs — CHANGES_REQUESTED** [review]
@fayazara I have tested the change and made some minor updates. The PR still has the following issues.

1. Resolve conversations action is not working with this change. 
2. There are some code which needs to be cleaned up, please see the individual changes.

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/AgentSelector.vue [hunk 10]
```diff
@@ -148,30 +159,30 @@ export default {
...
-  background-color: var(--white);
-  width: 75%;
+  box-shadow: var(--shadow-dropdown-pane);
+  max-width: 75%;
+  position: absolute;
+  right: var(--space-small);
+  top: 4.8rem;
+  transform-origin: top right;
+  width: auto;
+  z-index: 20;
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/AgentSelector.vue:173]:
```suggestion
  z-index: var(--z-index-twenty);
```

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/AgentSelector.vue [hunk 11]
```diff
@@ -194,6 +205,14 @@ export default {
       }
     }
   }
+  .triangle {
+    display: block;
+    z-index: 1;
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/AgentSelector.vue:210]: Use variables

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/UpdateActions.vue [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+  background-color: var(--white);
+  border-radius: var(--border-radius-large);
+  border: 1px solid var(--s-50);
+  box-shadow: var(--shadow-dropdown-pane);
+  position: absolute;
+  right: var(--space-small);
+  top: 48px;
+  transform-origin: top right;
+  width: auto;
+  z-index: 20;
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/UpdateActions.vue:137]:
```suggestion
  z-index: var(--z-index-twenty);
```

File: app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/UpdateActions.vue [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    padding-top: var(--space-zero);
+  }
+
+  .triangle {
+    display: block;
+    position: absolute;
+    right: 2.8rem;
+    text-align: left;
+    top: calc(var(--space-slab) * -1);
+    z-index: 1;
```
Comment (iamsivin) [inline app/javascript/dashboard/components/widgets/conversation/conversationBulkActions/UpdateActions.vue:158]: Use variables

File: app/javascript/shared/assets/stylesheets/z-index.scss [hunk 14]
```diff
@@ -2,6 +2,7 @@
   // z-index
   --z-index-minus: -1;
   --z-index-zero: 0;
+  --z-index-one: 0;
```
Comment (pranavrajs) [inline app/javascript/shared/assets/stylesheets/z-index.scss:5]: 1?

Comment (pranavrajs) [inline app/javascript/shared/assets/stylesheets/z-index.scss:5]:
```suggestion
  --z-index-one: 1;
```

**Review by pranavrajs — APPROVED** [review]
Looks good.

**Review by pranavrajs — APPROVED** [review]

**Review by iamsivin — APPROVED** [review]
Code LGTM 👍🏻

**Pull request merged**